fmpz_poly_roots/padic_roots.c

```diff
@@ -0,0 +1,148 @@
+#include <stdio.h>
+
+#include "padic.h"
+#include "padic_poly.h"
+#include "fmpz_poly_roots.h"
+
+
+void
+fmpz_poly_roots_padic_init2 (fmpz_poly_roots_padic_t roots, slong n)
+{
+  slong j;
+  
+  roots->x0 = flint_malloc (sizeof (padic_t) * n);
+  
+  for (j = 0; j < n; j++)
+    {
+      padic_init (roots->x0 + j);
+    }
+  
+  roots->multiplicity = flint_malloc (sizeof (slong) * n);
+  roots->num = n;
+  roots->alloc = n;
+}
+
+void
+fmpz_poly_roots_padic_clear (fmpz_poly_roots_padic_t roots)
+{
+  slong j;
+  
+  for (j = 0; j < roots->alloc; j++)
+    {
+      padic_clear (roots->x0 + j);
+    }
+  
+  flint_free (roots->x0);
+  flint_free (roots->multiplicity);
+}
+
+char*
+fmpz_poly_roots_padic_get_str (fmpz_poly_roots_padic_t roots, padic_ctx_t pctx)
+{
+  char * buffer = NULL;
+  size_t buffer_size = 0;
+  FILE *out = open_memstream(&buffer, &buffer_size);
+  
+  slong j;
+  
+  for (j = 0; j < roots->num; j++)
+    {
+      padic_fprint(out, roots->x0 + j, pctx);
+      flint_fprintf (out, " %wd\n", roots->multiplicity[j]);
+    }
+  
+  fclose(out);
+  
+  return buffer;
+}
+
+int
+fmpz_poly_roots_padic_fprint (FILE *file, fmpz_poly_roots_padic_t roots, padic_ctx_t pctx)
+{
+  slong j;
+  
+  for (j = 0; j < roots->num; j++)
+    {
+      padic_fprint(file, roots->x0 + j, pctx);
+      flint_fprintf (file, " %wd\n", roots->multiplicity[j]);
+    }
+
+  return 1;
+}
+
+int
+fmpz_poly_roots_padic_print (fmpz_poly_roots_padic_t roots, padic_ctx_t pctx)
+{
+  fmpz_poly_roots_padic_fprint(stdout, roots, pctx);
+  return 1;
+}
+
+static void
+padic_hensel_iteration (fmpz_poly_t poly, padic_t x, padic_ctx_t ctx)
+{
+  slong j;
+  padic_t tmp, y0, y1;
+  
+  padic_init (tmp);
+  padic_init (y0);
+  padic_init (y1);
+  
+  do
+    {
+      /* Horner evaluation of poly and poly' at x */
+      padic_set_fmpz (y0, poly->coeffs + poly->length - 1, ctx);
+      padic_zero (y1);
+      for (j = poly->length - 2; j >= 0; j--)
+	{
+	  padic_mul (y1, y1, x, ctx);
+	  padic_add (y1, y1, y0, ctx);
+	  padic_mul (y0, y0, x, ctx);
+	  padic_set_fmpz (tmp, poly->coeffs + j, ctx);
+	  padic_add (y0, y0, tmp, ctx);
+	}
+      /* Newton step: x -> x - poly / poly' */
+      padic_inv (y1, y1, ctx);
```

Comment:
I'd suggest a double iteration to compute rt and inv(f'(rt)) at the same time. Otherwise, this inv repeats the previous inve all the time.

Also, in case on many roots, a lifting of the factorisation is much faster than lifting of individual roots.

The precision should be dynamically increased using a lifting chain...

I guess most of this applies to the qadic case as well

Reply:
https://arxiv.org/pdf/1610.06837.pdf  
algo 13, steps 1-3

P. Panayi, Computation of Leopoldt’s p-adic regulator, Dissertation, University of East Anglia, 1995.
contains a "certified" p-adic root finder also in the hard cases